### PR TITLE
Use a custom schema for the search indexing metadata

### PR DESCRIPTION
**Pull request by Toflar**
Draft PR to fix https://github.com/contao/contao/issues/1236.

### CODE REVIEW

File: core-bundle/src/Resources/contao/pages/PageRegular.php [hunk 1]
```diff
@@ -823,24 +823,62 @@ protected function createFooterScripts($objLayout, $objPage = null)
...
+			$groups = array_map('intval', array_filter((array) $objPage->groups));
+			if (!empty($groups))
+			{
+				$audience[] = array
+				(
+					'@type' => 'PeopleAudience',
+					'name' => 'groups',
+					'identifier' => $groups,
+				);
+			}
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/pages/PageRegular.php]: this should only be applied if the page is `protected`

File: core-bundle/src/Resources/contao/pages/PageRegular.php [hunk 2]
```diff
@@ -823,24 +823,62 @@ protected function createFooterScripts($objLayout, $objPage = null)
...
 			if ($token !== null && $token->getUser() instanceof FrontendUser)
 			{
-				$meta['memberId'] = (int) $token->getUser()->id;
+				$audience[] = array
+				(
+					'@type' => 'PeopleAudience',
+					'name' => 'member',
+					'identifier' => (int) $token->getUser()->id,
+				);
+			}
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/pages/PageRegular.php]: Why do we override the group permissions with user permissions? There can never be group permissions then, because there's always a user logged in when viewing that information? I know that has been the case before (and this PR is only to rewrite the existing), but maybe you can explain?

**Review by aschempp — COMMENTED** [review]
I'm very unsure whether it's a good idea to _try_ to use schema.org for our internal data structure. The mapping is very limited and more "guessing" than actually knowing which meta properties we should use.
I would assume the issue with #1236 is not that we do use our own schema, but that the provided link does actually not point to a valid schema definition? So maybe we have to come up with JSON-LD schema information for our content? Which actually makes sense, so there's a "documentation" about the features?

**Comment by leofeyer** [comment]
I agree that we should use our own schema, so that search engines do not try to process our JSON-LD data.

**Comment by Toflar** [comment]
I don't think we should use our own schema. It's actually a big plus that other search engines can understand even more about our pages. Why wouldn't we want to have that? We already add quite a few schema.org tags using HTML, why not enhance it using JSON-LD?

**Comment by leofeyer** [comment]
Good point. I guess as long as the schema validates, it does not hurt to use the existing one.

**Comment by aschempp** [comment]
I've investigated the topic, and using our own schema is the only sensible approach to this.

Our own schema will allow **us** (someone reading Contao pages) to actually get meaningful values. **Only if** we use our own schema, a crawler can know with certainty that the current page is a Contao page. If it's a *generic* `schema.org/WebPage`, noone knows if `version = preview` is actually the preview version of a page (that must not be indexed). Another `schema.org/WebPage` could use `preview` for any other reason.

Using our own schema(s) actually gives much broader features: Our crawler could index the page, and it could (separately) index the news item. Search results could then actually show search results for news on the current website separately from the page search results. News can have a picture, and show up accordingly etc. all thanks to respective schema attributes.

To provide meaningful information to external crawlers, there are two options:
 1. Add **additional** schema information for `schema.org/WebPage`. Like generating similar JSON twice. That would be easy, and no problem for performance, as GZIP compresses duplicates anyway. I'm not sure if its' useful though, because we already have a lot of meta attributes that essentially generate the same information.
 2. Add a property to our own schema, that implements `schema.org/WebPage`. Our JSON would then basically contain a sub-key with the `schema.org/WebPage` stuff. I'm not sure if any parser (looking for a `@type WebPage`) would understand this, but it would certainly be valid JSON-LD.

At the end of the day, using `schema.org/WebPage` does not provide much benefit, because that information is already present on the page (using meta tags or `itemscope`/`itemtype`). Adding our own schema will allow our own crawler (or one understanding Contao) to read meaningful information.

**Comment by ausi** [comment]
I agree with @aschempp

I looked at the definitions of `potentialAction`, `conditionsOfAccess` and `audience`. I think they don’t fit our use case and would potentially break something if someone wants to use one of these properties for their intended use.

For example someone wants to add schema tags to the search form and adds the `SearchAction` as a `potentialAction` to the `WebPage` (see [schema.org/SearchAction](https://schema.org/SearchAction#website-1) Example 3). This would result in a conflict with our data.

IMO we should reuse the existing data that is already there for external crawlers:

* `language` => `<html lang="…">`
* `title` => `<title>…</title>`

And put the data that can only be meaningful to an internal crawler seperatly:

* `noSearch` => This would be a `<meta content="noindex">` but AFAIK we do want to differentiate between public searchable and private searchable. So we cannot use the data for external crawlers here.
* `protected` => There might be a standardized way to tell crawlers about protected pages, but `conditionsOfAccess` doesn’t seem right here IMO.
* `groups` => internal
* `fePreview` => internal
* `pageId` => internal
* `memberId` => internal

The internal data doesn’t have to be JSON-LD IMO, because it’s internal and cannot be used by someone without knowing about Contao.

**Comment by aschempp** [comment]
> The internal data doesn’t have to be JSON-LD IMO, because it’s internal and cannot be used by someone without knowing about Contao.

It should be JSON-LD, because Escargot / our crawler can understand it. That's the point of JSON-LD, to present machine-readable data to things crawling the page. But we need our own schema so one can **know** this is exactly what the data is. `@type: http://schema.org/WebPage` tells that `name: foobar` is the name of a webpage. Our type and props need to tell about our data.

**Comment by leofeyer** [comment]
As discussed in Mumble on February 13th, we want to have our own schema on schema.contao.org and retrieve the data as suggested by @ausi [above](https://github.com/contao/contao/pull/1250#issuecomment-583641228).

**Comment by Toflar** [comment]
After looking into this, our discussed solution doesn't work so I had to mark it up for discussion again. 

It seems like https://search.google.com/structured-data/testing-tool is not really a proper JSON LD validator but rather just validates `schema.org` and Google specific JSON LD data.
You can validate that by just copying the very first example of https://json-ld.org/ into the testing tool. It will tell you that `There was an invalid type in your JSON-LD.` which is not true, it's a perfectly valid type, it's just one Google doesn't know and doesn't care about.

So it seems like we cannot and don't have to get rid of this error.

**Comment by Toflar** [comment]
I've still updated the schema URL so that we could (or still can) introduce the correct schema and adjusted the current behaviour to read the language and title from HTML rather than the JSON-LD context as suggested by @ausi.

File: core-bundle/src/Resources/contao/pages/PageRegular.php [hunk 3]
```diff
@@ -825,24 +825,15 @@ protected function createFooterScripts($objLayout, $objPage = null)
...
 				'protected' => (bool) $objPage->protected,
 				'groups' => array_map('intval', array_filter((array) $objPage->groups)),
 				'fePreview' => System::getContainer()->get('contao.security.token_checker')->isPreviewMode()
 			);
 
-			$token = System::getContainer()->get('security.token_storage')->getToken();
-
-			if ($token !== null && $token->getUser() instanceof FrontendUser)
-			{
-				$meta['memberId'] = (int) $token->getUser()->id;
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/pages/PageRegular.php:843]: Do we not need the member ID anymore?

Comment (Toflar) [reply core-bundle/src/Resources/contao/pages/PageRegular.php:843]: We never did, that was just a „maybe we can use it one day“ :)

Comment (leofeyer) [reply core-bundle/src/Resources/contao/pages/PageRegular.php:843]: I see. 😄

**Comment by leofeyer** [comment]
I also ran some tests and the following JSON+LD is valid according to the Google test tool:

```json
{
  "@context": "https://schema.org",
  "@type": "Page",
  "pageId": 10,
  "noSearch": false,
  "protected": false,
  "groups": [],
  "fePreview": false
}
```

I'm pretty sure that this is just a coincidence, but as we will be using an "invalid" schema anyway, why not use one that does not trigger validation errors?

**Comment by Toflar** [comment]
But it‘s incorrect. Just because it validates, doesn‘t mean it‘s correct. Imho what Google does is wrong.

**Comment by leofeyer** [comment]
Yes, I agree.

What about using good old `<meta>` tags instead of JSON-LD?

```html
<meta name="page_id" content="10">
<meta name="no_search" content="">
<meta name="protected" content="">
<meta name="groups" content="">
<meta name="fe_preview" content="">
```

One advantage would be that we could entirely omit tags without content.

**Comment by Toflar** [comment]
I don't understand what you are trying to achieve, really. JSON-LD was invented for exactly the use case we're having here. We don't need to bend over backwards just to please any tool that validates incorrectly.

File: core-bundle/src/Search/Document.php [hunk 4]
```diff
@@ -76,6 +81,15 @@ public function getBody(): string
         return $this->body;
     }
 
+    public function getContentCrawler(): Crawler
```
Comment (leofeyer) [inline core-bundle/src/Search/Document.php:84]: Why did you add this method?

Comment (Toflar) [reply core-bundle/src/Search/Document.php:84]: To be able to fetch the content crawler? So we don't need multiple instances of the DomCrawler for everybody who wants to crawl the content.

**Review by leofeyer — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.
